source/extensions/filters/http/grpc_web/grpc_web_filter.cc

```diff
@@ -143,27 +239,57 @@ Http::FilterDataStatus GrpcWebFilter::decodeData(Buffer::Instance& data, bool en
 }
 
 // Implements StreamEncoderFilter.
-Http::FilterHeadersStatus GrpcWebFilter::encodeHeaders(Http::ResponseHeaderMap& headers, bool) {
+Http::FilterHeadersStatus GrpcWebFilter::encodeHeaders(Http::ResponseHeaderMap& headers,
+                                                       bool end_stream) {
   if (!is_grpc_web_request_) {
     return Http::FilterHeadersStatus::Continue;
   }
 
   if (doStatTracking()) {
     chargeStat(headers);
   }
+
+  needs_transformation_for_non_proto_encoded_response_ =
+      needsTransformationForNonProtoEncodedResponse(headers, end_stream);
+
   if (is_text_response_) {
     headers.setReferenceContentType(Http::Headers::get().ContentTypeValues.GrpcWebTextProto);
   } else {
     headers.setReferenceContentType(Http::Headers::get().ContentTypeValues.GrpcWebProto);
   }
-  return Http::FilterHeadersStatus::Continue;
+
+  if (end_stream || !needs_transformation_for_non_proto_encoded_response_) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  response_headers_ = &headers;
+  return Http::FilterHeadersStatus::StopIteration;
 }
 
-Http::FilterDataStatus GrpcWebFilter::encodeData(Buffer::Instance& data, bool) {
+Http::FilterDataStatus GrpcWebFilter::encodeData(Buffer::Instance& data, bool end_stream) {
   if (!is_grpc_web_request_) {
     return Http::FilterDataStatus::Continue;
   }
 
+  // When the upstream response (this is also relevant for local reply, since gRPC-Web request is
+  // not a gRPC request which makes the local reply's is_grpc_request set to false) is not a gRPC
+  // response, we set the "grpc-message" header with the upstream body content.
+  if (needs_transformation_for_non_proto_encoded_response_) {
+    const auto* encoding_buffer = encoder_callbacks_->encodingBuffer();
+    if (!end_stream) {
+      if (encoding_buffer != nullptr &&
```

Comment:
Maybe leave a comment for why we want to limit buffering here in addition to when we construct grpc-message

Reply:
Added.